// @redwoodjs/web/dist/components/FatalErrorBoundary.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _jsxRuntime = require("react/jsx-runtime");
class InfallibleErrorBoundary extends _react.default.Component {
  constructor() {
    super(...arguments);
    this.state = {
      hasError: false
    };
  }
  static getDerivedStateFromError() {
    return {
      hasError: true
    };
  }
  render() {
    if (this.state.hasError) {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)("h1", {
        children: "Something went wrong and we are unable to show this page."
      });
    }
    return this.props.children;
  }
}
class FatalErrorBoundary extends _react.default.Component {
  constructor() {
    super(...arguments);
    this.state = {
      hasError: false,
      error: undefined
    };
  }
  static getDerivedStateFromError(error) {
    return {
      hasError: true,
      error
    };
  }
  render() {
    const {
      page: Page
    } = this.props;
    if (this.state.hasError) {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)(InfallibleErrorBoundary, {
        children: /*#__PURE__*/(0, _jsxRuntime.jsx)(Page, {
          error: this.state.error
        })
      });
    }
    return this.props.children;
  }
}
var _default = exports.default = FatalErrorBoundary;
